clippy_lints/src/large_enum_variant.rs

```diff
@@ -113,23 +114,35 @@ impl<'tcx> LateLintPass<'tcx> for LargeEnumVariant {
                                 def.variants[(second.1).0].span,
                                 &format!("and the second-largest variant is {} bytes:", second.0),
                             );
-                            if variant.fields.len() == 1 {
-                                let span = match def.variants[i].data {
-                                    VariantData::Struct(fields, ..) | VariantData::Tuple(fields, ..) => {
-                                        fields[0].ty.span
-                                    },
-                                    VariantData::Unit(..) => unreachable!(),
-                                };
-                                if let Some(snip) = snippet_opt(cx, span) {
-                                    diag.span_suggestion(
-                                        span,
-                                        help_text,
-                                        format!("Box<{}>", snip),
-                                        Applicability::MaybeIncorrect,
-                                    );
-                                    return;
+
+                            let mut sugg: Vec<(Span, String)> = vec![];
+                            let fields = match def.variants[i].data {
+                                VariantData::Struct(fields, ..) | VariantData::Tuple(fields, ..) => fields,
+                                VariantData::Unit(..) => unreachable!(),
+                            };
+                            let mut fields_size = Vec::new();
+                            for (ind, val) in variant.fields.iter().enumerate() {
+                                let ty = cx.tcx.type_of(val.did);
+                                if let Some(size_info) = cx.layout_of(ty).ok().map(|l| l.size.bytes()) {
+                                    fields_size.push((ind, size_info));
                                 }
                             }
+                            fields_size.sort_by(|a, b| (a.1.cmp(&b.1)));
```

Comment:
This is mostly repeating what has been done earlier in the function. Can we combine some of the logic? For instance, if this sorted Vec is available earlier, then the largest and second largest variants are simply the first two elements. Also there is too much tuple field access `.1` and structs would help.

Reply:
> This is mostly repeating what has been done earlier in the function. Can we combine some of the logic? For instance, if this sorted Vec is available earlier, then the largest and second largest variants are simply the first two elements. Also there is too much tuple field access `.1` and structs would help.

Hi, thanks for reviewing.
For example:
enum LargeEnum8 {
     VariantOk(i32, u32),
     ContainingMoreThanOneField([i32; 8000], [i32; 2], [i32; 9500], [i32; 30]),
}
The first calculation is for the size of VariantOk and ContainingMoreThanOneField, and the second is calculating size for the fields of ContainingMoreThanOneField, such as [i32; 8000], [i32; 2]...
Because the fields of ContainingMoreThanOneField only need to be calculated under the condition of difference > maximum_size_difference_allowed, I tend to separate them.

Reply:
The first loop does get the layout and size of every field of every variant, so there isn't any work being saved. You could have `VariantInfo { size: usize, fields: Vec<FieldInfo> }`, create a `Vec<VariantInfo>`, sort it by size, and then only sort `fields` for the largest variant.

Reply:
> The first loop does get the layout and size of every field of every variant, so there isn't any work being saved. You could have `VariantInfo { size: usize, fields: Vec<FieldInfo> }`, create a `Vec<VariantInfo>`, sort it by size, and then only sort `fields` for the largest variant.

Yes. You are right, I got it wrong. fixed